pkg/reconciler/resolver/pipelinerun_resolver_reconciler.go

```diff
@@ -0,0 +1,122 @@
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/reconciler"
+)
+
+type PipelineRunResolverReconciler struct {
+	// Implements reconciler.LeaderAware
+	reconciler.LeaderAwareFuncs
+
+	kubeClientSet     kubernetes.Interface
+	pipelineClientSet clientset.Interface
+	pipelinerunLister listers.PipelineRunLister
+	configStore       reconciler.ConfigStore
+}
+
+var _ controller.Reconciler = &PipelineRunResolverReconciler{}
+
+func (r *PipelineRunResolverReconciler) Reconcile(ctx context.Context, key string) error {
+	if r.configStore != nil {
+		ctx = r.configStore.ToContext(ctx)
+	}
+
+	logger := logging.FromContext(ctx)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return fmt.Errorf("invalid resource key %q: %v", key, err)
+	}
+
+	pr, err := r.pipelinerunLister.PipelineRuns(namespace).Get(name)
+	if err != nil {
+		return fmt.Errorf("error getting pipelinerun %s/%s: %w", namespace, name, err)
+	}
+
+	req := resolution.PipelineRunResolutionRequest{
+		KubeClientSet:     r.kubeClientSet,
+		PipelineClientSet: r.pipelineClientSet,
+		PipelineRun:       pr,
+	}
+
+	// TODO(sbws): At some point I'd like to add a new status reason. Something
+	// like Succeeded/Unknown with reason:PipelineResolving. This would improve on the
+	// existing situation where PipelineRuns are stored with a blank status until
+	// bundle / cluster resolution is completed. For now, though, that does not match
+	// behaviour with the existing reconcilers so it is not added yet.
+
+	err = req.Resolve(ctx)
+	switch {
+	case pr == nil:
+		// The pipelinerun we were told to resolve doesn't exist, therefore
+		// we can't update it with a failure condition.
+		// TODO(sbws): Does this match our reconciler's current behaviour?
+		return controller.NewPermanentError(err)
+	case err == resolution.ErrorPipelineRunAlreadyResolved:
+		// Nothing to do: another process has already resolved the pipelinerun.
+		return nil
+	case err != nil:
+		reason := resolution.ReasonPipelineRunResolutionFailed
+		if e, ok := err.(*resolution.Error); ok {
+			reason = e.Reason
+		}
+		pr.Status.MarkFailed(reason, err.Error())
+		logger.Error(err)
+		return err
+	}
+
+	resolution.CopyPipelineMetaToPipelineRun(req.ResolvedPipelineMeta, pr)
+
+	pr.Status.PipelineSpec = req.ResolvedPipelineSpec
+
+	// Update() because we want any labels and annotations from the pipeline's
+	// meta persisted in the PipelineRun. Matches expectations/behaviour of the
+	// current PipelineRun reconciler.
+	newPR, err := r.pipelineClientSet.TektonV1beta1().PipelineRuns(pr.Namespace).Update(ctx, pr, metav1.UpdateOptions{})
+	if err != nil {
+		err = fmt.Errorf("error committing updated pipeline spec to pipelinerun %s/%s: %w", namespace, name, err)
+		pr.Status.MarkFailed(resolution.ReasonPipelineRunResolutionFailed, err.Error())
+		logger.Error(err)
+		return fmt.Errorf("error updating pipelinerun spec: %w", err)
+	}
```

Comment:
Any reason to do an update instead of a patch ? I think it could this result in a weird race condition (if there is a mutating webhook doing something with `PipelineRun` in the meantime).

Reply:
Very interesting! I hadn't thought about the mutating webhook case.  I'll try this with a patch operation instead.

Reply:
As far as I can tell our test clients don't support `Patch()` operations on resources. I certainly update the reconcilers to use `Patch` instead of `Update` but the unit tests will rely on a workaround that uses `Update` still. This means we're not exercising the exact same resolution behaviour between unit tests and release. I'm hopeful integration will work without any such workarounds though.  @vdemeester are you ok with this tradeoff?

Reply:
I've updated the `pkg/reconciler/taskrun/taskrun_test.go` file so that the tests pass. Changes aren't perfect yet (in particular - it looks like the number of events emitted has changed between `main` and my branch) but it shows [how the resolver updates in unit tests use `Update()` while the resolver reconciler uses `Patch()`](https://github.com/tektoncd/pipeline/pull/4108/files#diff-8babe88e4c33b69be0e453cdec72b43387c85f6b560347345cb91a44505ba977R3459-R3490)